Fix to #225, added better error reporting for internal compiler errors.

inJuly0 opened the pull request:
Previously, when an incorrect table was passed into a function dealing with a certain type of table, only an "impossible" error was thrown followed by a Lua stack trace.

Now it's possible to see the tag which caused the error, and wherever appropriate, an error message following it. Calls to `error("impossible")` have been replaced with a call to `typedecl.tag_error(tag, "(optional error message.)")`.

Hopefully, this should fix [#225 ](https://github.com/pallene-lang/pallene/issues/225).

hugomg commented on pallene/types.lua:
```diff
@@ -246,8 +251,10 @@ function types.tostring(t)
     elseif tag == "types.T.Record" then
         return t.name
+    elseif not typedecl.tag_matches(tag, "types.T") then
+        typedecl.tag_error(tag, "not a type.")
```
I also think we might not need this elseif

hugomg commented on pallene/checker.lua:
```diff
@@ -455,7 +455,7 @@ function Checker:check_stat(stat)
                 stat.step = ast.Exp.Float(stat.limit.loc, 1.0)
             else
-                error("impossible")
+                typedecl.tag_error(loop_type._tag, "incorrect loop step type.")
```
Nice. This is exactly the kind of situation that benefits the most from a custom error message.

Small suggestion: what do you think about changing the error message to `loop type is not a number type`?

inJuly0 replied:
Yeah, this error message is a bit unclear. fixed.

hugomg commented on pallene/checker.lua:
```diff
@@ -687,7 +689,7 @@ function Checker:check_exp_synthesize(exp)
             exp._type = types.T.Boolean()
         else
-            error("impossible")
+            typedecl.tag_error(tag, string.format("unknown unary operator '%s'.", op))
```
With the new default error message I proposed, I think we could use `typedecl.tag_error(op)`.

hugomg commented on pallene/checker.lua:
```diff
@@ -797,7 +799,7 @@ function Checker:check_exp_synthesize(exp)
 
         else
-            error("impossible")
+            typedecl.tag_error(tag, string.format("unknown binary operator '%s'.", op))
```
Again, the error message should reference `op`, not the `tag`. With the error message I suggested I think we could use `typedecl.tag_error(op)`.

hugomg commented on pallene/coder.lua:
```diff
@@ -39,7 +39,9 @@ local function ctype(typ)
     elseif tag == "types.T.Any"      then return "TValue"
-    else error("impossible")
+    elseif typedecl.tag_is_type(tag) then
+        typedecl.tag_error(tag, "unable to get corresponding c type.")
```
I think the default error message is fine here.

hugomg commented on pallene/coder.lua:
```diff
@@ -117,8 +119,12 @@ local function set_stack_slot(typ, dst_slot, value)
     elseif tag == "types.T.Any"      then tmpl = "setobj(L, $dst, &$src);"
-    else error("impossible")
+    elseif not typedecl.tag_is_type(tag) then
+        typedecl.tag_error(tag, "not a type.")
```
This is another case where I think the default error message is ok.

hugomg commented on pallene/coder.lua:
```diff
@@ -165,8 +171,8 @@ local function pallene_type_tag(typ)
     elseif tag == "types.T.Record"   then return "LUA_TUSERDATA"
-    elseif tag == "types.T.Any"    then error("value is not a tag")
-    else error("impossible")
+    elseif tag == "types.T.Any"      then typedecl.tag_error(tag, "value is not a tag")
```
I just noticed that this error message is outdated. In the past, the `any` type used to be called `value` but we forgot to update the error message when that changes.

Could you please change the error message to `"Any" is not a Lua type tag`?

inJuly0 replied:
will do

hugomg commented on pallene/coder.lua:
```diff
@@ -324,8 +330,10 @@ function Coder:c_value(value)
+        typedecl.tag_error(tag, "unable to get C expression for this value type.")
     else
-        error("impossible")
+        typedecl.tag_error(tag, "not a value.")
```
I think it is ok if we use the default error message here.

hugomg commented on pallene/to_ir.lua:
```diff
@@ -495,10 +497,10 @@ function ToIR:exp_to_value(cmds, exp, _recursive)
 
             elseif cname._tag == "checker.Name.Builtin" then
-                error("not implemented")
+                typedecl.tag_error(cname._tag, "not implemented")
```
If the error happens, it should be clear what the type tag was because of the preceding if statement. I think we can continue using `error("not implemented")` here.

hugomg commented on pallene/to_ir.lua:
```diff
@@ -775,7 +780,9 @@ function ToIR:exp_to_assignment(cmds, dst, exp)
-                error("impossible")
+                typedecl.tag_error(tag,
+                    string.format("error casting from type '%s' to '%s'",
+                        types.tostring(src_typ), types.tostring(dst_typ)))
```
I think the `tag_error(tag` is not appropriate here, because the tags that actually matter for the error are `dst_typ._tag` and `src_typ.tag` and not the `exp._tag`.

I like the new error message but I think we could just use a plain `error(string.format(` instead.

hugomg commented on pallene/typedecl.lua:
```diff
@@ -82,4 +82,35 @@ function typedecl.declare(module, mod_name, type_name, constructors)
+function typedecl.tag_error(tag, message)
+    local error_msg = message
+        and string.format("error at tag '%s': %s", tag, message)
+        or string.format("error at tag '%s'.", tag)
```
What do you think about changing the error message to not mention the word "tag". This way it would also be usable in other kinds of if-elseif like the binary operators and the builtin function names. Suggestion:

```
function typedecl.tag_error(tag, message)
    message = message or "input has the wrong type or an elseif case is missing"
    error(string.format("the if-elseif did not handle case %s: %s", tag, message))
end
```

inJuly0 replied:
Yes, I think that would allow for more uses of `tag_error`.
My reasoning was that the function name implies the error is always associated with a tag.

However I do agree this allows for more liberal use of the function. I'll change it. :+1:

hugomg commented on pallene/to_ir.lua:
```diff
@@ -649,7 +651,7 @@ function ToIR:exp_to_assignment(cmds, dst, exp)
                 table.insert(cmds, ir.Cmd.BuiltinTostring(loc, dsts, xs))
             else
-                error("impossible")
+                typedecl.tag_error(cname._tag, "unknown builtin function.")
```
This tag_error should be referencing the `bname`, not the `cname._tag`.

```
typedecl.tag_error(bname)
```

hugomg reviewed (changes requested):
Lovely! Thanks for going through this. I think you even found some cases that I would have missed if I tried doing this myself.

I do have some suggestion for changes before we can merge this though.

The first is that if we choose a good default error message for tag_error then I think we should be able to use the default error message in more places. I think it would be good if we had a clear rule for whether we should have a special error message or not. That way, we don't need to decide for every possible if-elseif whether it should have a special error message or not. One rule of thumb we could follow is that if the if-elseif covers all the possible cases then it can use the default error message and if it only covers a subset of the possible cases then it should have a special error message.

Another thing that we could change is that there are some places in the code that could be updated to use  the new `tag_matches` function. You can find them by running `grep -R 'string.match(.*tag'`.

inJuly0 commented on pallene/typedecl.lua:
```diff
@@ -82,4 +82,35 @@ function typedecl.declare(module, mod_name, type_name, constructors)
+--
+-- @param tag: The type name (string)
+-- @param tag_prefix: The prefix to test (string)
+
```
Ah, I somehow missed that. will do.

inJuly0 commented on pallene/typedecl.lua:
```diff
@@ -82,4 +82,35 @@ function typedecl.declare(module, mod_name, type_name, constructors)
+
+function typedecl.tag_matches(tag, tag_prefix)
+    return type(tag) == "string"
+        and string.find(tag, tag_prefix, 1, true) ~= nil
```
Oh shoot, I forgot about that too.

 Yeah, I definitely have intentions of making contributions in the future since I love compilers and have been exploring type systems as of late.

I actually messed this up because I was coding on my Linux PC, If I manage to get Pallene built up and running on windows these silly errors most likely won't persist.

inJuly0 commented on pallene/types.lua:
```diff
@@ -194,7 +198,8 @@ local function equivalent(t1, t2, is_gradual)
     else
-        return error("impossible")
+        return typedecl.tag_error(tag1,
+            string.format("attempt to check equivalence of types %s and %s.", tag1, tag2))
```
Oh, I noticed that in places where `string.matches` is used, the prefix being tested is `"types.T."`.
I think it would make more sense to check every tag prefix with the the second "." included, no ?

inJuly0 commented:
I made the changes as asked, excuse the CI fails in older commits, I made some mistakes by merging into master too early. Everything should now work as expected.

One part I couldn't figure out is this line, in coder.lua:1508 :

```lua
local name = assert(string.match(cmd._tag, "^ir%.Cmd%.(.*)$"))
```

I had originally replaced that with this:

```lua
local name = cmd._tag
assert(typedecl.tag_matches(name, "ir.Cmd."))
```
however, doing so would cause the tests to fail, and I seem to be unable to diagnose why. 
So I left that one line unchanged. It would be awesome if you could explain that. cheers.

hugomg commented:
> string.match(cmd._tag, "^ir%.Cmd%.(.*)$")
> however, doing so would cause the tests to fail, and I seem to be unable to diagnose why.

The `string.match()` here is using a string capture. For example, if `cmd._tag` is `ir.Cmd.Binop` then `name` will be `Binop`.

Maybe we should change the tag_matches function to return this string instead of just true/false?

hugomg reviewed (changes requested):
Thanks. I marked all the things that you fixed as resolved. There are only a handful that were missed that are still there:

- The tag_error for unknown unary operators
- There is stll an `assert(string.match(typ._tag, "^types%.T%."))` in types.lua
- There is the `name = assert(string.match(cmd._tag, "^ir%.Cmd%.(.*)$"))` that broke the CI

hugomg commented:
About the broken CI, if we really wanted we could use `git rebase -i` to fix the old commits so they pass the CI. However, in this case it is only a very small problem so I don't think we have to bother with it.

inJuly0 commented:
> The `string.match()` here is using a string capture. For example, if `cmd._tag` is `ir.Cmd.Binop` then `name` will be `Binop`.
> 
> Maybe we should change the tag_matches function to return this string instead of just true/false?

Yeah that would seem like a good idea. 
I missed the capture group, kinda rusty with regex. Thanks for explaining.

I feel like the name `tag_matches` implies a boolean. What would you say about letting that one line stay as is instead of changing intent of the function for just one usage ? Thoughts ?

inJuly0 commented:
> Good idea. It would be safer to always add the second .. What do you think about adding this . inside the tag_matches function, so the caller doesn't have to remember to add it themselves?

Since the `tag_matches` function does prefix matching, I personally would find it easier if it only matched against the string it receives as an argument instead of adding a "`.`". It could, at times cause confusion. Thoughts ?

hugomg commented:
>  I personally would find it easier if it only matched against the string it receives as an argument instead of adding a ".".

Maybe it should have a different name instead of `tag_matches` then? What we we are trying to see is if a certain tag (such as `"types.T.Boolean"`) has a given type (such as `types.T`). We aren't trying to match against an arbitrary string...

> I feel like the name tag_matches implies a boolean. 

If you can think of a good name that doesn't imply a boolean we could use that. If not, I think it is OK if we continue using `string.match("ir%.Cmd%.(.*)")` in that one place.

inJuly0 commented:
> If you can think of a good name that doesn't imply a boolean we could use that.

I went with `match_tag` and made the changes as asked. 
The function now adds a second "." and captures and returns everything following it.

Probably not the best name but this is loosely similar to what `string.match` does in Lua , or Javascript,
match a pattern in a string and return the matched sequence. Except this returns the part following the `.`.

Do let me know if I missed something yet again :D

hugomg commented on pallene/typedecl.lua:
```diff
@@ -82,4 +82,31 @@ function typedecl.declare(module, mod_name, type_name, constructors)
+-- @param tag_prefix: The prefix to test (string)
+function typedecl.match_tag(tag, tag_prefix)
+    return type(tag) == "string" and string.match(tag, tag_prefix .. ".(.*)")
+end
```
One problem with using string.match here is that special characters in the tag_prefix will be treated as a regular expression. For example, a prefix of `types.T` would also match `typesaT` or `typesbT`, which is not what we want to happen.

One possible fix is to use string.sub to get the parts of the string that we are looking for:

```

function typedecl.match_tag(tag, tag_prefix)
    local n = #tag_prefix
    if type(tag) == "string" and
       string.sub(tag, 1, n) == tag_prefix and
       string.byte(tag, n+1) == 46 -- "."
    then
        return string.sub(tag, n+2)
    else
        return false
    end
end
```

hugomg reviewed (changes requested):
Nice! I like the new name.

I found a couple of small bugs but they should be easy to fix and after that I think this should be good for merging.

It would also be nice if we added some test cases to typedecl_spec.lua. In particular, we should test that something like `typedecl.match_tag("abc.d.e", "a.c")` returns `false` because that is a bug that already hit us before and it happened again today. Would you be able to help by writing those tests?

inJuly0 commented on pallene/print_ir.lua:
```diff
@@ -214,7 +215,8 @@ local function Cmd(cmd)
     else
-        local tagname = assert(string.match(cmd._tag, "^ir.Cmd.(.*)"))
+        local tagname = cmd._tag
+        assert(typedecl.match_tag(tagname, "ir.Cmd"))
```
Oh yes, my bad.

inJuly0 commented:
> It would also be nice if we added some test cases to typedecl_spec.lua. In particular, we should test that something like typedecl.match_tag("abc.d.e", "a.c") returns false
Yeah, I should be able to put together some tests for that.

inJuly0 commented:
I updated that `assert` guard, changed `match_tag` to no longer treat it's arguments as regex strings.
Added 3 basic test cases to `typedecl_spec.lua`. 

Do let me know if I messed up somewhere. :D

hugomg commented on spec/typedecl_spec.lua:
```diff
@@ -8,4 +8,10 @@ describe("Typedecl", function()
+        assert.falsy(typedecl.match_tag("foo.Bar.baz", "f.o.Bar"))
+        assert.truthy(typedecl.match_tag("foo.Bar.baz", "foo.Bar"))
+        assert.falsy(typedecl.match_tag("types.T.Float", "types.T."))
+    end, "")
```
* We could put each test case in a separate `it` block. That way we can give a descriptive name for each of them. For example the first test can say that it is testing that the "." is not being interpreted as a regex character.
* The second test case could test that the return value is `"baz"` instead of only testing that it is truthy
* I think we don't need to put that `"")` at the end

inJuly0 replied:
True, should have gone with more descriptive test logs.
Done :+1:.

hugomg reviewed (changes requested):
Just some small things in the typedecl_spec and then I think this is done.

hugomg reviewed (approved):
Perfect! Thanks a lot for the contribution! It's much appreciated.
